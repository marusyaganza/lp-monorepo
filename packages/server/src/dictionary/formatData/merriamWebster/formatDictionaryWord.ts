import { DictionaryWord, Language } from '../../../generated/graphql';
import { uniq } from 'lodash';

import {
  filterString,
  getAudioUrl,
  getDefs,
  getDefFromRef,
  getImgUrl,
  formatDictionaryEntity,
  formatHw
} from './helpers';

import { MWDictionaryWord } from './types/types';

export function formatDictionaryWord(
  word: MWDictionaryWord
): DictionaryWord | undefined {
  const { def, meta, hwi, art, shortdef, cxs, suppl } = word;
  const { uuid, id, stems, offensive, lang } = meta;
  const isSpanish = lang == 'es';
  const name = isSpanish && hwi?.hw ? hwi.hw : filterString(id);
  const metaData = {
    uuid,
    stems,
    name,
    isOffensive: offensive
  };
  const defs = Array.isArray(def) ? getDefs(def) : getDefFromRef(cxs);
  const pronunciation = hwi?.prs?.[0];
  const transcription = pronunciation?.mw || formatHw(hwi?.hw);
  const audio = pronunciation?.sound?.audio;
  const audioUrl = getAudioUrl(audio, lang);
  const language = isSpanish ? Language.Spanish : Language.English;
  const imgUrl = getImgUrl(art?.artid);
  const imgDesc = art?.capt ? formatDictionaryEntity(art?.capt) : null;
  const particle = word?.fl ?? 'noun';
  const shortDef = shortdef?.length ? shortdef : defs.map(def => def.def);
  const conjugation = suppl?.cjts?.length ? suppl.cjts : null;

  if (!defs?.length) {
    return;
  }

  return {
    ...metaData,
    particle,
    defs,
    transcription,
    imgUrl,
    imgDesc,
    audioUrl,
    shortDef: uniq(shortDef),
    conjugation,
    language
  };
}
