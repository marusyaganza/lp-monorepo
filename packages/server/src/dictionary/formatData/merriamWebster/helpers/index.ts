export * from './general/general';
export * from './resource/resource';
export * from './string/string';
export * from './def/def';
