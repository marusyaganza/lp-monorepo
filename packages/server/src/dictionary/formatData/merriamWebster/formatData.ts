import { DictionaryWord, SearchResult } from '../../../generated/graphql';
import { isMWDictionaryWord } from './types/typeGuards';
import {
  assertIsTypedArray,
  isDictionaryWord,
  isString,
  isTypedArray
} from '../../../types/typeGuards';
import { MWDictionaryWord } from './types/types';
import { formatDictionaryWord } from './formatDictionaryWord';

export function formatData(data: unknown): SearchResult[] {
  if (isTypedArray<MWDictionaryWord>(data, isMWDictionaryWord)) {
    const result = data
      .map(entry => formatDictionaryWord(entry))
      .filter(Boolean);
    assertIsTypedArray<DictionaryWord>(result, isDictionaryWord);
    return result;
  }
  if (isTypedArray<string>(data, isString)) {
    return [{ suggestions: data }];
  }
  return [];
}
