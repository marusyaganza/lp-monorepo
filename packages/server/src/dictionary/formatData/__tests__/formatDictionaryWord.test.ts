import murtherWord from '../../fetchWord/mockFetchWord/mocks/ENGLISH/murther';
import heartWord from '../../fetchWord/mockFetchWord/mocks/ENGLISH/heart';
import caerWord from '../../fetchWord/mockFetchWord/mocks/SPANISH/caerse';
import idiomaWord from '../../fetchWord/mockFetchWord/mocks/SPANISH/idioma';
import { formatDictionaryWord } from '../merriamWebster/formatDictionaryWord';

const incompleteWord = {
  meta: {
    id: 'voluminous',
    uuid: '0d01b967-971f-4ec5-8fe0-10513d29c39b'
  },
  hwi: {
    hw: 'vo*lu*mi*nous'
  },
  fl: 'adjective',
  shortdef: [
    'having or marked by great volume or bulk : large; also : full',
    'numerous',
    'filling or capable of filling a large volume or several volumes'
  ]
};

describe('formatDictionaryWord', () => {
  test('should return correct result with Spanish verb', async () => {
    const result = formatDictionaryWord(caerWord[0]);
    expect(result?.conjugation).toHaveLength(20);
    expect(result).toMatchSnapshot();
  });

  test('should return correct result with Spanish noun', async () => {
    const result = formatDictionaryWord(idiomaWord[0]);
    expect(result).toMatchSnapshot();
  });

  test('should return correct result minimal English word', async () => {
    // @ts-expect-error: testing with incomplete data
    const result = formatDictionaryWord(murtherWord[0]);
    expect(result).toMatchSnapshot();
  });

  test('should return correct result with complex English word', async () => {
    // @ts-expect-error: testing with incomplete data
    const result = formatDictionaryWord(heartWord[0]);
    expect(result).toMatchSnapshot();
  });

  test('should return undefined with incomplete data', async () => {
    // @ts-expect-error: testing with incomplete data
    const result = formatDictionaryWord(incompleteWord);
    expect(result).toBe(undefined);
  });
});
