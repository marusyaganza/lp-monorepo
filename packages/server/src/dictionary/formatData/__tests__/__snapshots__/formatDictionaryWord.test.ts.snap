// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatDictionaryWord should return correct result minimal English word 1`] = `
{
  "audioUrl": "mock_audio_endpoint/en/us/mp3/m/murthe01.mp3",
  "conjugation": null,
  "defs": [
    {
      "def": "chiefly dialectal variant of <i>murder</i>",
    },
  ],
  "imgDesc": null,
  "imgUrl": undefined,
  "isOffensive": false,
  "language": "ENGLISH",
  "name": "murther",
  "particle": "noun",
  "shortDef": [
    "chiefly dialectal variant of <i>murder</i>",
  ],
  "stems": [
    "murder",
    "murther",
    "murthered",
    "murthering",
    "murthers",
  ],
  "transcription": "ˈmər-t͟hər",
  "uuid": "87f2219c-6fcb-4390-9e2a-427384ef13a3",
}
`;

exports[`formatDictionaryWord should return correct result with Spanish noun 1`] = `
{
  "audioUrl": "mock_audio_endpoint/es/me/mp3/i/idiom01sp.mp3",
  "conjugation": null,
  "defs": [
    {
      "def": "language",
      "examples": [
        {
          "text": "el idioma inglés",
          "translation": "the English language",
        },
      ],
    },
  ],
  "imgDesc": null,
  "imgUrl": undefined,
  "isOffensive": false,
  "language": "SPANISH",
  "name": "idioma",
  "particle": "masculine noun",
  "shortDef": [
    "language",
  ],
  "stems": [
    "idioma",
  ],
  "transcription": "idioma",
  "uuid": "63882a81-032e-4dc1-9c22-bb683cb1bad9",
}
`;

exports[`formatDictionaryWord should return correct result with Spanish verb 1`] = `
{
  "audioUrl": "mock_audio_endpoint/es/me/mp3/c/caer001sp.mp3",
  "conjugation": [
    {
      "cjfs": [
        "cayendo",
        "caído",
      ],
      "cjid": "gppt",
    },
    {
      "cjfs": [
        "caigo",
        "caes",
        "cae",
        "caemos",
        "caéis",
        "caen",
      ],
      "cjid": "pind",
    },
    {
      "cjfs": [
        "caía",
        "caías",
        "caía",
        "caíamos",
        "caíais",
        "caían",
      ],
      "cjid": "pret",
    },
    {
      "cjfs": [
        "caí",
        "caíste",
        "cayó",
        "caímos",
        "caísteis",
        "cayeron",
      ],
      "cjid": "pprf",
    },
    {
      "cjfs": [
        "caeré",
        "caerás",
        "caerá",
        "caeremos",
        "caeréis",
        "caerán",
      ],
      "cjid": "futr",
    },
    {
      "cjfs": [
        "caería",
        "caerías",
        "caería",
        "caeríamos",
        "caeríais",
        "caerían",
      ],
      "cjid": "cond",
    },
    {
      "cjfs": [
        "caiga",
        "caigas",
        "caiga",
        "caigamos",
        "caigáis",
        "caigan",
      ],
      "cjid": "psub",
    },
    {
      "cjfs": [
        "cayera",
        "cayeras",
        "cayera",
        "cayéramos",
        "cayerais",
        "cayeran",
      ],
      "cjid": "pisb1",
    },
    {
      "cjfs": [
        "cayese",
        "cayeses",
        "cayese",
        "cayésemos",
        "cayeseis",
        "cayesen",
      ],
      "cjid": "pisb2",
    },
    {
      "cjfs": [
        "cayere",
        "cayeres",
        "cayere",
        "cayéremos",
        "cayereis",
        "cayeren",
      ],
      "cjid": "fsub",
    },
    {
      "cjfs": [
        "he caído",
        "has caído",
        "ha caído",
        "hemos caído",
        "habéis caído",
        "han caído",
      ],
      "cjid": "ppci",
    },
    {
      "cjfs": [
        "había caído",
        "habías caído",
        "había caído",
        "habíamos caído",
        "habíais caído",
        "habían caído",
      ],
      "cjid": "ppsi",
    },
    {
      "cjfs": [
        "hube caído",
        "hubiste caído",
        "hubo caído",
        "hubimos caído",
        "hubisteis caído",
        "hubieron caído",
      ],
      "cjid": "pant",
    },
    {
      "cjfs": [
        "habré caído",
        "habrás caído",
        "habrá caído",
        "habremos caído",
        "habréis caído",
        "habrán caído",
      ],
      "cjid": "fpin",
    },
    {
      "cjfs": [
        "habría caído",
        "habrías caído",
        "habría caído",
        "habríamos caído",
        "habríais caído",
        "habrían caído",
      ],
      "cjid": "cpef",
    },
    {
      "cjfs": [
        "haya caído",
        "hayas caído",
        "haya caído",
        "hayamos caído",
        "hayáis caído",
        "hayan caído",
      ],
      "cjid": "ppfs",
    },
    {
      "cjfs": [
        "hubiera caído",
        "hubieras caído",
        "hubiera caído",
        "hubiéramos caído",
        "hubierais caído",
        "hubieran caído",
      ],
      "cjid": "ppss1",
    },
    {
      "cjfs": [
        "hubiese caído",
        "hubieses caído",
        "hubiese caído",
        "hubiésemos caído",
        "hubieseis caído",
        "hubiesen caído",
      ],
      "cjid": "ppss2",
    },
    {
      "cjfs": [
        "hubiere caído",
        "hubieres caído",
        "hubiere caído",
        "hubiéremos caído",
        "hubiereis caído",
        "hubieren caído",
      ],
      "cjid": "fpsb",
    },
    {
      "cjfs": [
        "-",
        "cae",
        "caiga",
        "caigamos",
        "caed",
        "caigan",
      ],
      "cjid": "impf",
    },
  ],
  "defs": [
    {
      "def": "to fall",
      "examples": [
        {
          "text": "cayó al suelo",
          "translation": "he fell on the floor/ground",
        },
        {
          "text": "lo dejó caer",
          "translation": "she dropped it",
        },
      ],
    },
    {
      "def": "to drop away, to slope",
    },
    {
      "def": "to fall (of night)",
    },
    {
      "def": "to collapse, to fall",
    },
    {
      "def": "to hang (down)",
    },
    {
      "def": "to realize, to understand",
      "examples": [
        {
          "text": "caer (en) que …",
          "translation": "to realize that …",
        },
      ],
    },
    {
      "def": "to fall into (a trap, etc.)",
      "examples": [
        {
          "text": "caer en el error de",
          "translation": "to make the mistake of",
        },
        {
          "text": "caer en manos de",
          "translation": "to fall into the hands of",
        },
        {
          "text": "caer en la tentación",
          "translation": "to give in to temptation",
        },
      ],
    },
    {
      "def": "to fall out of favor",
    },
    {
      "def": "to fall ill",
    },
    {
      "def": "to sit well/poorly with someone",
      "examples": [
        {
          "text": "me caes bien",
          "translation": "I like you",
        },
      ],
    },
  ],
  "imgDesc": null,
  "imgUrl": undefined,
  "isOffensive": false,
  "language": "SPANISH",
  "name": "caer",
  "particle": "intransitive verb",
  "shortDef": [
    "to fall",
    "to drop away, to slope",
    "to fall (of night)",
  ],
  "stems": [
    "caer",
    "caer en",
    "caer en desgracia",
    "caer enfermo",
    "caerle bien/mal a alguien",
    "caerle bien a alguien",
    "caerle mal a alguien",
    "caerse",
  ],
  "transcription": "caer",
  "uuid": "aac40ca2-64b3-4515-8a3d-91f11127a5b1",
}
`;

exports[`formatDictionaryWord should return correct result with complex English word 1`] = `
{
  "audioUrl": "mock_audio_endpoint/en/us/mp3/h/heart001.mp3",
  "conjugation": null,
  "defs": [
    {
      "def": "something resembling a heart in shape",
    },
    {
      "def": "a game in which the object is to avoid taking tricks see trick containing hearts",
    },
    {
      "def": "intellect",
    },
    {
      "def": "see also take heart",
      "examples": [
        {
          "text": "never lost <i>heart</i>",
          "translation": undefined,
        },
      ],
    },
    {
      "def": "one's innermost character, feelings, or inclinations",
      "examples": [
        {
          "text": "knew it in his <i>heart</i>",
          "translation": undefined,
        },
        {
          "text": "a man after my own <i>heart</i>",
          "translation": undefined,
        },
      ],
    },
    {
      "def": "the younger central compact part of a leafy rosette (such as a head of lettuce or stalk of celery)",
    },
  ],
  "imgDesc": "heart 1a: <i>1</i> aorta, <i>2</i> pulmonary artery, <i>3</i> left atrium, <i>4</i> left ventricle, <i>5</i> right ventricle, <i>6</i> right atrium",
  "imgUrl": "mock_img_endpoint/heart.gif",
  "isOffensive": false,
  "language": "ENGLISH",
  "name": "heart",
  "particle": "noun",
  "shortDef": [
    "a hollow muscular organ of vertebrate animals that by its rhythmic contraction acts as a force pump maintaining the circulation of the blood",
    "a structure in an invertebrate animal functionally analogous to the vertebrate heart",
    "breast, bosom",
  ],
  "stems": [
    "at heart",
    "by heart",
    "heart",
    "hearts",
    "to heart",
  ],
  "transcription": "ˈhärt",
  "uuid": "f0b7bdb0-3194-4621-8410-992ac81192f1",
}
`;
